Add blog post on practical agents for real-world development (#9485)
Microsoft/vscode-docs: blogs/2026/03/05/making-agents-practical-for-real-world-development.md

From enforcing policies with hooks to guiding agents mid-response, validating UI features with integrated browser tools, and bringing structured skills directly into the editor, this release focuses on making agents reliable collaborators for real development work.

## Give agents the right context

Long-running development work depends on maintaining continuity. As sessions grow, these updates make the current experience more effective for complex and long-running sessions, reducing the need to repeatedly restate project details.

### Handle large outputs

Large diffs, generated files, or extensive logs can overwhelm a session if treated as an inline context.

VS Code now manages large outputs by streaming them to temporary files and prioritizing the most relevant information for the model. This keeps agents focused on the right details while optimizing context usage without additional work.

<video src="embedded-terminal-streaming.mp4" title="Video showing terminal output auto-expanding for long-running commands." autoplay muted controls></video>

### Share agent memory

> Agent memory now spans coding agents, CLI workflows, and code review interactions, allowing agents to share context and build on common knowledge across sessions.

Reviewer: ```suggestion
Agents in Visual Studio Code use memory to retain relevant context. Agent memory now spans coding agents, CLI workflows, and code review interactions.

Rather than starting from scratch each session, agents recall your preferences, apply lessons from previous tasks, and build up knowledge about your codebase over time.
```

Reviewer: Link to the docs: https://code.visualstudio.com/docs/copilot/agents/memory

Architectural decisions, naming conventions, and prior refactors remain part of the conversation, so you spend less time restating intent and more time continuing the work.

### Compact long sessions

> As conversations expand, VS Code automatically compacts older history. Earlier discussion is summarized, key decisions are preserved, and space is freed for ongoing work.

Reviewer: ```suggestion
As conversations expand, VS Code automatically compacts older history. Earlier discussions are summarized, key decisions are preserved, and space is freed up for ongoing work.
```

> You can also compact the session manually using `/compact`, providing instructions about what context should be preserved or discarded as the conversation evolves.

Reviewer: ```suggestion
Previously, you had no control over when context compaction was happening and what information was retained after compacting. Maybe you discussed several implementation variants, and only one specific one is important to remember and build upon.

Now you can manually run context compaction for a session by typing `/compact`. And in doing so, you can give the agent additional instructions on what information to keep or discard.
```

> This keeps sessions focused without interrupting the workflow.

Reviewer: ```suggestion
Especially when working with long-running sessions and dealing with lots of context, controlling compaction keeps the agent focused on the key information.
```

![Screenshot showing the context window control and the compact option.](context-compaction.png)

## Agent controls

As agents take on more responsibility, the way you interact with them matters just as much as what they generate. These updates make it easier to control the conversation and guide outcomes during active work.

### Guide the agent while it works

Agents sometimes head down the wrong path, and you can usually tell before they finish.

Previously, you had to wait for a response to complete it before redirecting it. Now, you can intervene while the agent is generating a response, guiding the direction of the work without restarting or losing context.

> You can also send follow-up instructions to refine or redirect the task as the work progresses.

Reviewer: ```suggestion
And if think of some extra tasks the agent should perform, you can now queue follow-on requests for the agent to perform once it has finished it's current task. If you queue up multiple requests, you can easily change the order in which they need to be performed.
```
